// demux-eos/dist/MongoBlock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class MongoBlock {
    constructor(rawBlock) {
        this.actions = this.collectActionsFromBlock(rawBlock);
        this.blockInfo = {
            blockNumber: rawBlock.block_num,
            blockHash: rawBlock.block_id,
            previousBlockHash: rawBlock.block.previous,
            timestamp: new Date(rawBlock.block.timestamp),
        };
    }
    collectActionsFromBlock(rawBlock = { actions: [] }) {
        return this.flattenArray(rawBlock.block.transactions.map(({ trx }) => {
            return trx.transaction.actions.map((action, actionIndex) => {
                // Delete unneeded hex data if we have deserialized data
                if (action.payload) {
                    delete action.hex_data; // eslint-disable-line
                }
                return {
                    type: `${action.account}::${action.name}`,
                    payload: Object.assign({ transactionId: trx.id, actionIndex }, action),
                };
            });
        }));
    }
    flattenArray(arr) {
        return arr.reduce((flat, toFlatten) => flat.concat(Array.isArray(toFlatten) ? this.flattenArray(toFlatten) : toFlatten), []);
    }
}
exports.MongoBlock = MongoBlock;
